gson/src/main/java/com/google/gson/internal/bind/RecordTypeAdapterFactory.java

```diff
@@ -0,0 +1,480 @@
+            // https://stackoverflow.com/a/67127067
+            Constructor<T> recordConstructor = (Constructor<T>) type.getRawType().getConstructor(componentTypes);
+
+            RecordField[] recordFields = new RecordField[recordComponents.length];
+            for (int i = 0; i < recordComponents.length; i++) {
+                // We need the field for the Gson Excluder, so that we can correctly determine if a field should
+                // be included or not.
+                Field field = RECORD_HELPER.getField(recordComponents[i], type);
+                recordFields[i] = recordFieldFactory.createRecordField(field, recordComponents[i]);
+            }
+            return new RecordTypeAdapterImpl<>(recordFields, recordConstructor);
+        }
+        catch (NoSuchMethodException e) {
+            // We hit this either because we do not find the expected constructor, or we do not find a method
+            // name that matches the field name. In either of these classes, this is not a record, and we
+            // do not support serialization of it. The contract for TypeAdapterFactory is to return null for
+            // unsupported types. Gson will then attempt other means of creating a TypeAdapter. If all fails, an
+            // exception will be created where Gson was invoked to serialize/deserialize a type.
+            return null;
+        }
+    }
+
+    /**
+     * Factory to construct our own {@link RecordField} instances. This requires some reflection into the internal
+     * of Gson, which has logic to handle the {@link JsonAdapter} annotation. We also need to reflect into Gson
+     * to fetch the {@link Excluder}, that handles the {@link com.google.gson.annotations.Expose},
+     * {@link com.google.gson.annotations.Since} and {@link com.google.gson.annotations.Until} annotations. This
+     * allows us to closely follow the Gson mechanisms wrt. serialization/deserialization.
+     */
+    private static final class RecordFieldFactory {
+
+
+        private final Gson gson;
+        private final Excluder excluder;
+        private final JsonAdapterAnnotationTypeAdapterFactory jsonAdapterFactory;
+        private final ConstructorConstructor constructorConstructor;
+
+        public RecordFieldFactory(Gson gson, Excluder excluder,
+                JsonAdapterAnnotationTypeAdapterFactory jsonAdapterFactory,
+                ConstructorConstructor constructorConstructor) {
+            this.gson = gson;
+            this.excluder = excluder;
+            this.jsonAdapterFactory = jsonAdapterFactory;
+            this.constructorConstructor = constructorConstructor;
+        }
+
+        private RecordField createRecordField(Field field, Object recordComponent) {
+            TypeToken<?> fieldType = TypeToken.get(RECORD_HELPER.getGenericType(recordComponent));
+            JsonAdapter jsonAdapter = RECORD_HELPER.getAnnotation(recordComponent, JsonAdapter.class);
```

Comment:
Similar to https://github.com/google/gson/pull/2197/files#r978712348, I think the annotation has to be obtained from private field instead of the component.